### PR TITLE
 Move line parsing out of FileReadingIterator

### PR DESCRIPTION
**Pull request by mfussenegger**
The `FileReadingIterator` was not only reading lines from URIs, but also
parsing the contents via `InputRow` and a list of `LineCollectorExpression`

Because of the `LineCollectorExpression` it had very tight coupling with:

- LineProcessor
- LineParser
- LineContext

This moves the parsing functionality out of the `FileReadingIterator`
into the `LineProcessor` for more loose coupling.

The `LineProcessor` is now itself a `BatchIterator<Row>`, using a
`BatchIterator<LineCursor>` as source and merging parts of `LineParser`
into it. Neither of these components could be used meaningful in a
standalone way.

Note that this is just an intermediate step. The `LineProcessor` as it
is in this PR probably won't survive. To get rid of the `_raw` import
mechanism I plan to merge it somewhat with the `IndexWriterProjector`.

This also deletes some benchmarks. I plan to re-add new one's once the
full picture is in place

### CODE REVIEW

File: server/src/test/java/io/crate/execution/engine/collect/files/LineParserTest.java [hunk 1]
```diff
@@ -1,107 +0,0 @@
...
-import java.io.StringReader;
-import java.net.URI;
-import java.util.List;
-
-import org.junit.jupiter.api.Test;
-
-import io.crate.analyze.CopyFromParserProperties;
-import io.crate.execution.dsl.phases.FileUriCollectPhase;
-
-class LineParserTest {
```
Comment (mfussenegger) [inline server/src/test/java/io/crate/execution/engine/collect/files/LineParserTest.java:38]: Deleted these as I'm now re-using the SkippingBatchIterator. Added a new `test_file_collect_source_returns_iterator_that_can_skip_lines` test case to cover that.

File: server/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java [hunk 2]
```diff
@@ -273,17 +344,17 @@ private String getLine(BufferedReader reader) throws IOException {
...
         killed = BatchIterator.CLOSED;
         backOffPolicy.forEachRemaining((delay) -> {});
     }
 
-    private void releaseBatchIteratorState() {
+    private void reset() {
         fileInputsIterator = null;
         currentInputUriIterator = null;
         currentInput = null;
-        currentUri = null;
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java:286]: don't we need smth like `cursor.uri = null`?

File: server/src/main/java/io/crate/execution/engine/collect/files/LineProcessor.java [hunk 3]
```diff
@@ -21,60 +21,124 @@
...
+                    lineContext.rawSource(json);
+                } catch (Throwable parseError) {
+                    lineContext.setCurrentParsingFailure(parseError.getMessage());
+                }
+                return true;
+            }
+            return false;
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/execution/engine/collect/files/LineProcessor.java:132]: afaik, this BatchIterator comes after `FileReadingIterator` ie this won't go into `lineContext.LineCursor.failure` (and it shouldn't)

I remember CSV parsing library throws IOExceptions on parsing (I think it's csvReader.readValues doing that) - ie we probably should qualify this as a parsing failure and call `lineContext.setCurrentParsingFailure`?

File: server/src/main/java/io/crate/execution/engine/collect/sources/FileCollectSource.java [hunk 4]
```diff
@@ -85,22 +88,28 @@ public CompletableFuture<BatchIterator<Row>> getIterator(TransactionContext txnC
...
+            fileUriCollectPhase.nodeIds().size(),
+            getReaderNumber(fileUriCollectPhase.nodeIds(), clusterService.state().nodes().getLocalNodeId()),
+            fileUriCollectPhase.withClauseOptions(),
+            threadPool.scheduler()
+        );
+        CopyFromParserProperties parserProperties = fileUriCollectPhase.parserProperties();
+        LineProcessor lineProcessor = new LineProcessor(
+            parserProperties.skipNumLines() > 0
+                ? new SkippingBatchIterator<>(fileReadingIterator, (int) parserProperties.skipNumLines())
+                : fileReadingIterator,
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/execution/engine/collect/sources/FileCollectSource.java:105]: I like this!

**Review by BaurzhanSakhariev — COMMENTED** [review]
LGTM in general, added some minor comments and questions. 

Correct me if I'm wrong but looks like this code can be BWC (if we do Version dependant `CopyFromPlan`). We still can collect _raw and specific columns for <5.5 Versions it's just collected in different place (in LineProcessor now).

It's out of the scope for now but Im wondering what's the plan in general, will we support rolling upgrades for COPY FORM?

File: server/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java [hunk 5]
```diff
@@ -50,98 +50,171 @@
...
+
+    private final LineCursor cursor;
     private final ScheduledExecutorService scheduler;
     private final Iterator<TimeValue> backOffPolicy;
 
+    public static class LineCursor {
+        private URI uri;
+        private long lineNumber;
+        private String line;
+        private IOException failure;
```
[the remark this answers is not in the document]
Comment (mfussenegger) [reply server/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java:114]: I think the question is if there's a good reason to do it early.
It's a lossy conversation. `IOException` carries more information than `String`. I don't see a benefit in throwing away the information early.

**Review by mfussenegger — COMMENTED** [review]
> Correct me if I'm wrong but looks like this code can be BWC (if we do Version dependant CopyFromPlan). We still can collect _raw and specific columns for <5.5 Versions it's just collected in different place (in LineProcessor now).

> It's out of the scope for now but Im wondering what's the plan in general, will we support rolling upgrades for COPY FORM?

The current PR is just shuffling things around without any streaming or plan implications.

I'm still not sure what the final streaming implications will be.

**Review by BaurzhanSakhariev — APPROVED** [review]
Thanks!

**Pull request merged**